### PR TITLE
Acceleration of chacha20 on aarch64 by SVE

### PR DESCRIPTION
**Pull request by daniel-hu-arm**
This patch accelerates chacha20 on aarch64 when Scalable Vector Extension
(SVE) is supported by CPU. Tested on modern micro-architecture with
256-bit SVE, it has the potential to improve performance up to 20%.
Optimization is only applied to the scenario where input data is over
256 bytes

Test result:
With SVE
type            1024 bytes   8192 bytes  16384 bytes
ChaCha20        1596373.33k  1639167.32k  1643173.21k

Without SVE (by Neon/Scalar)
type            1024 bytes   8192 bytes  16384 bytes
chacha20        1355487.91k  1372678.83k  1372662.44k

The assembly code has been reviewed internally by
ARM engineer Fangming.Fang@arm.com

Signed-off-by: Daniel Hu <Daniel.Hu@arm.com>

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/arm64cpuid.pl
Comment (paulidale) [inline crypto/arm64cpuid.pl:111]: Is `eor` correct here?  If so it would be assigning zero wouldn't it?  Which would be kind of pointless.

**Review by paulidale — COMMENTED** [review]
Any chance of an independent review of the assembly file?

File: crypto/arm64cpuid.pl
Comment (daniel-hu-arm) [reply crypto/arm64cpuid.pl:111]: yes, the only purpose of this instruction is to test if SVE feature is supported by CPU (z0 register only available with SVE). if not, the instruction will cause SIGKILL sent to user process. if yes, it will set a scratch register (z0) to zero with no side-effect to caller.

We have the assembly code internally reviewed in Arm. but not sure if we can find reviewer from other perspective.

Comment (t8m) [reply crypto/arm64cpuid.pl:111]: One thing that could help would be if you (or someone from Arm) were willing to become community maintainer for one of the aarch64 configuration targets that are currently unsupported:
https://www.openssl.org/policies/platformpolicy.html

Comment (daniel-hu-arm) [reply crypto/arm64cpuid.pl:111]: @t8m we are keen to support the community. can you let us know what you expect of the role? in terms of people, equipment, process, effort, etc.

Comment (t8m) [reply crypto/arm64cpuid.pl:111]: As you can see from the description of what the Community support means on the page linked above the requirements are fairly minimal - basically it means the person that is the community contact will need to response to platform-specific defects that are found either by CI or the community in some reasonable time.
Currently the CI is using GitHub Actions their runners which means x86_64 architecture. However we run there CI jobs which use cross-compiler for aarch64 and also run some tests with qemu emulation.

Eventually we would like to be able to run an internal CI based on buildbot and support external buildbot workers which could enable you to use some aarch64 machine(s) hosted by you as CI. But we are not there yet.

Comment (daniel-hu-arm) [reply crypto/arm64cpuid.pl:111]: @t8m thanks for explanation. It is good to know you are running some tests in your CI using qemu emulation.

Internally we run CI on our aarch64 hardware and qemu as well, and as you said, potentially we could help to run standard openssl CI task in the future when it is technically feasible.

We will have some discussion in our team, and see if we can identify person(s) to interface the community. I'll get back to you later. 

Do you have a list of priority of these unsupported aarch64 configurations? (Linux or Android aarch64 could be the priority, I think)

Comment (t8m) [reply crypto/arm64cpuid.pl:111]: It would be really up to you to choose which one you're able to support. And of course you can support also multiple ones.

Comment (zorrorffm) [reply crypto/arm64cpuid.pl:111]: @t8m, could we identify 4 arm engineers contributing to OpenSSL as community maintainers?  We expect there are some backups and supports from different time zone. Besides individual identity is provided, maybe an email alias like openssl-arm-support.arm.com can be set up to reach all of us.  Is there any formal process or guide we can follow to apply for being the community maintainer?

Another question about platform policy. Reading https://www.openssl.org/blog/blog/2021/09/22/OpenSSL3-fips-submission,  there are some platforms that were tested before OpenSSL 3.0 FIPS module was submitted for FIPS validation, are these tested platforms chosen from primary and community platforms?  If not, how a platform can be involved to be tested when validate FIPS module? 

Thanks.

Comment (t8m) [reply crypto/arm64cpuid.pl:111]: You can have multiple people as the contact persons for the platform.

Then one of them simply has to make a PR against the openssl/web repo to make the necessary modifications to the platform policy page and have that PR approved. Once approved and merged its official.

There are no more requirements than what I wrote above in this comment thread.

The FIPS module validation is a separate topic - perhaps @mattcaswell can expand on that.

Comment (mattcaswell) [reply crypto/arm64cpuid.pl:111]: > are these tested platforms chosen from primary and community platforms

Actually no. The FIPS work was sponsored by a number of companies (see https://www.openssl.org/blog/blog/2018/09/25/fips/). We included platforms nominated by our sponsors as well as a couple that the project itself nominated.

Moving forward the OpenSSL Management Committee (OMC) is responsible for maintaining the validation. It is likely that the OMC will want to keep the list of platforms tested to some minimal set. Our primary and secondary lists (currently no platforms are secondary) are likely to influence decisions about any additional platforms to be added if any.

Comment (zorrorffm) [reply crypto/arm64cpuid.pl:111]: Thanks @t8m for the guide, we will make that PR to apply for the community maintainer.   And thanks @mattcaswell for the explanation about FIPS platform.

**Review by paulidale — APPROVED** [review]

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (samuel-lee-msft) [inline crypto/chacha/asm/chacha-armv8-sve.pl:181]: In the presence of SVE2, wouldn't XAR be preferable here?

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (samuel-lee-msft) [inline crypto/chacha/asm/chacha-armv8-sve.pl:188]: Any reason not to use SRI here?
https://developer.arm.com/documentation/ddi0602/2022-03/SVE-Instructions/SRI--Shift-right-and-insert--immediate--?lang=en

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (daniel-hu-arm) [reply crypto/chacha/asm/chacha-armv8-sve.pl:181]: yes, we had the suspicion xar could be better off, but since we don't have any sve2 hardware to verify, we decided to defer sve2 specific optimization to future.

the code has been written in a way to use sve2 instruction for some scenario. the assumption is that we will have to revisit the code for sve2 on future hardware, at which point we will optimize and formally enable sve2 for chacha20.

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (samuel-lee-msft) [inline crypto/chacha/asm/chacha-armv8-sve.pl:564]: Is there actually any benefit to having 2 versions of the inner block (one predicated and one unpredicated)?

It makes sense that you would only use the predicated parts for combining with plain/ciphertext, but it seems like needless complication and codesize to optionally predicate the INNER_BLOCK logic.

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (samuel-lee-msft) [inline crypto/chacha/asm/chacha-armv8-sve.pl:502]: This seems like it probably shouldn't stay here for merging to OpenSSL unless you're saying it improves performance on real hardware, which would be very odd.

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (daniel-hu-arm) [reply crypto/chacha/asm/chacha-armv8-sve.pl:564]: as demonstrated on latest micro-architecture, predicate instruction has significant overhead (10%+). It means if SVE register is saturated (or every element utilized), we are better off using unpredicated instruction. 

some instruction for inner-block (e.g. XAR etc.) will be unpredicated anyway, but other instructions like EOR, ADD, etc. need to avoid predication if it is not necessary.

File: crypto/arm64cpuid.pl
Comment (samuel-lee-msft) [inline crypto/arm64cpuid.pl:111]: Judging by #18086 this, consider using .inst directive?

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (daniel-hu-arm) [reply crypto/chacha/asm/chacha-armv8-sve.pl:502]: yes, it does improve performance on real hardware, and yes, it is very odd.

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (samuel-lee-msft) [reply crypto/chacha/asm/chacha-armv8-sve.pl:564]: I'm saying why not _only_ have the unpredicated version for INNER_BLOCK?

File: crypto/arm64cpuid.pl
Comment (daniel-hu-arm) [reply crypto/arm64cpuid.pl:111]: it looks like there are many similar instances to fix in the same file if we need to switch to .inst altogether. 

but i will change the sve probe in this case.

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (samuel-lee-msft) [reply crypto/chacha/asm/chacha-armv8-sve.pl:564]: Agree that the decision should be made based on data.

I suspect that for V1 there is no real benefit to predication for 5/6/7 blocks - if you have 5 or more active .S lanes I would expect each SVE instruction needs to be split into 2 128b micro-operations (uops), and this would have the same impact on the backend as an unpredicated instruction (if anything probably be a bit slower).

There could be a benefit when:
(block_count % 8) is in the range [1,4] (i.e. 9...12, 17...24, etc.)
so in the final iteration when the predicated version is used, each SVE instruction only needs to produce 1 128b uop. It's not clear to me from the V1 SWOG that this actually happens.

Of course using predication in the tail case may show a bigger improvement for larger vector lengths depending on how the microarchitecture is set up (i.e. if predication can reduce uop count). In the long term I would expect that most of the time 1 SVE instruction will be 1 uop regardless of predication.

Comment (daniel-hu-arm) [reply crypto/chacha/asm/chacha-armv8-sve.pl:564]: let me run some test for the scenario where (block_count %8) in the range of [1, 4], it is probably easier to see the impact from test.

for foreseeable future, perhaps SVE vector length will mostly be 128/256 bits, (the only thing bigger that i am aware of, is a Fujitsu super computer on 512 bits). At the moment, we are only able to access 256-bit V1 hardware for testing.

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (daniel-hu-arm) [reply crypto/chacha/asm/chacha-armv8-sve.pl:260]: thanks for the nice suggestion, much cleaner.

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (samuel-lee-msft) [inline crypto/chacha/asm/chacha-armv8-sve.pl:607]: So here it looks like if input is not a multiple of block_size you are copying the final partial input block to the stack, en/decrypting it in place, and then copying it back to the output.

I do not think this is necessary. If instead you rounded $blocks up to the nearest whole block (i.e. if $len&63 is non-zero, increment $blocks), and then predicated the load/stores in `output_regs` byte-wise ($dlen would have to be calculated from $len, rather than $blocks), then partial blocks would be handled in the main loop.
This should also be substantially faster for unusual-sized inputs.

If I'm missing something which makes the byte-wise predication in the main loop infeasible, then at least be sure to wipe the data copied to the stack when you copy out from it (otherwise we're potentially leaving sensitive data on the stack).

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (daniel-hu-arm) [reply crypto/chacha/asm/chacha-armv8-sve.pl:564]: i think easier way is just to always handle small blocks or tail (that is, smaller than SVE lanes) in Neon. SVE will only handle full size. from performance point of view, perhaps this is the main scenario that matters, and could make code simpler.

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (daniel-hu-arm) [reply crypto/chacha/asm/chacha-armv8-sve.pl:607]: i'll let the Neon handle the tail. SVE will no longer process tail.

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (samuel-lee-msft) [reply crypto/chacha/asm/chacha-armv8-sve.pl:564]: Yeah, I suppose that approach might work out better in practice, at least until SVE vector lengths are normally >256b - the ASIMD code handling tail cases needs to be in the OpenSSL binary anyway.
There is some elegance to having SVE predication handle the unexpected buffer lengths cleanly at the point of loading, xor-ing the buffer with the state, and storing, but it might not be enough faster to be worth the additional complexity.

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (samuel-lee-msft) [inline crypto/chacha/asm/chacha-armv8-sve.pl:270]: Small optimization opportunity:
I notice currently z13-z15 are not being used and SVE codepath is only used when vector length >= 256b.

You could load a copy of state (64 bytes fits into 2x256b vectors) into z13, z14 once at the start of SVE code, and in these places you want to dup the state (add_states and SVE_LOAD_STATES), use [DUP (indexed)](https://developer.arm.com/documentation/ddi0602/2022-03/SVE-Instructions/DUP--indexed---Broadcast-indexed-element-to-vector--unpredicated--?lang=en), rather than loading into scalars and duplicating from scalar to vector registers.

Similarly, you could only use the INDEX instruction once at the beginning, writing to z15, and update it with [INCW (vector)](https://developer.arm.com/documentation/ddi0602/2022-03/SVE-Instructions/INCD--INCH--INCW--vector---Increment-vector-by-multiple-of-predicate-constraint-element-count-?lang=en), rather than moving the counter between scalar and vector registers.

**Comment by samuel-lee-msft** [comment]
With the simplification to remove handling tail cases in SVE I think the code is a lot cleaner - LGTM overall

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (daniel-hu-arm) [reply crypto/chacha/asm/chacha-armv8-sve.pl:271]: changes to INDEX makes code slightly faster, also the code looks cleaner, so I pushed the new changes.

Comment (samuel-lee-msft) [reply crypto/chacha/asm/chacha-armv8-sve.pl:271]: I'm guessing you also already tried [LD1RW](https://developer.arm.com/documentation/ddi0602/2022-03/SVE-Instructions/LD1RW--Load-and-broadcast-unsigned-word-to-vector-?lang=en) in place of LDP (scalar) + DUP (scalar), or DUP (indexed)?

Comment (daniel-hu-arm) [reply crypto/chacha/asm/chacha-armv8-sve.pl:271]: ld1rw will slow down the code. it seems the mix of scalar load and DUP are the best combination so far.

File: crypto/chacha/asm/chacha-armv8-sve.pl
Comment (daniel-hu-arm) [reply crypto/chacha/asm/chacha-armv8-sve.pl]: tried everything i could think of, this weird code still needed.

Comment (samuel-lee-msft) [reply crypto/chacha/asm/chacha-armv8-sve.pl]: Fair enough!

**Review by paulidale — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by paulidale** [comment]
Merged, thanks for the contribution.

**Pull request closed without merging**
